File: foundations-jdbc-dsl-scala/src/scala/dev/typr/foundations/scala/DeleteBuilder.scala
package dev.typr.foundations.scala

import dev.typr.foundations.Fragment
import dev.typr.foundations.dsl

import java.sql.Connection
import _root_.scala.jdk.CollectionConverters.*
import _root_.scala.jdk.OptionConverters.*
import scala.util.Using

class DeleteBuilder[Fields, Row](private val javaBuilder: dsl.DeleteBuilder[Fields, Row]) {

  def where(predicate: Fields => SqlExpr[Boolean]): DeleteBuilder[Fields, Row] = {
    new DeleteBuilder(javaBuilder.where((fields: Fields) => SqlExpr.toJavaBool(predicate(fields))))
  }

  def execute(using connection: Connection): Int = {
    javaBuilder.execute(connection)
  }

  def executeReturning(parser: ResultSetParser[List[Row]])(using connection: Connection): List[Row] = {
    sql() match {
      case Some(fragment) =>
        // SQL implementation - execute directly with Scala parser
        val query = fragment.append(Fragment.lit(" RETURNING *"))
        Using.resource(connection.prepareStatement(query.render())) { ps =>
          query.set(ps)
          Using.resource(ps.executeQuery()) { rs =>
            parser(rs)
          }
        }
      case None =>
        // Mock implementation - parser is ignored, just convert result
        javaBuilder.executeReturning(connection, null).asScala.toList
    }
  }

  def sql(): Option[Fragment] = {
    javaBuilder.sql().toScala
  }
}

object DeleteBuilder {
  def of[Fields, Row](
      tableName: String,
      structure: RelationStructure[Fields, Row],
      dialect: dsl.Dialect
  ): DeleteBuilder[Fields, Row] = {
    new DeleteBuilder(dsl.DeleteBuilder.of(tableName, structure, dialect))
  }
}

File: typr-dsl-kotlin/src/kotlin/dev/typr/dslkt/DeleteBuilder.kt
package dev.typr.dslkt

import dev.typr.dsl.DeleteBuilder as JavaDeleteBuilder
import dev.typr.dsl.Structure
import dev.typr.dsl.Dialect
import dev.typr.foundationskt.Fragment
import dev.typr.foundationskt.Connection

/**
 * Kotlin facade for SQL DELETE queries with type-safe operations.
 * Delegates to the Java implementation while providing Kotlin-friendly APIs.
 */
class DeleteBuilder<Fields, Row> internal constructor(
    private val javaBuilder: JavaDeleteBuilder<Fields, Row>
) {

    /**
     * Add a WHERE clause to the delete.
     * Consecutive calls will be combined with AND.
     */
    fun where(predicate: SqlExprExtensions.(Fields) -> SqlExpr<Boolean>): DeleteBuilder<Fields, Row> {
        return DeleteBuilder(javaBuilder.where { fields -> SqlExprExtensionsInstance.predicate(fields).underlying })
    }

    /**
     * Execute the delete and return the number of affected rows.
     */
    fun execute(connection: Connection): Int {
        return javaBuilder.execute(connection.javaConnection)
    }

    /**
     * Execute the delete and return the deleted rows (using RETURNING clause).
     */
    fun executeReturning(connection: Connection, parser: dev.typr.foundationskt.ResultSetParser<List<Row>>): List<Row> {
        return javaBuilder.executeReturning(connection.javaConnection, parser.underlying).toList()
    }

    /**
     * Get the SQL for debugging purposes. Returns null if backed by a mock repository.
     */
    fun sql(): Fragment? {
        return javaBuilder.sql().map { Fragment(it) }.orElse(null)
    }

    companion object {
        /**
         * Create a DeleteBuilder for a table.
         */
        fun <Fields, Row> of(
            tableName: String,
            structure: RelationStructure<Fields, Row>,
            dialect: Dialect
        ): DeleteBuilder<Fields, Row> {
            return DeleteBuilder(JavaDeleteBuilder.of(tableName, structure, dialect))
        }
    }
}
